### PR TITLE
[memory64] Keep type of memory.size and memory.grow on copy

### PR DESCRIPTION
**Pull request by backes**
When copying a MemorySize or MemoryGrow instruction (e.g. for inlining),
transfer the memory type also to the copy. Otherwise it will always be
i32, even if memory64 should be used.

This fixes issue #4530.

### CODE REVIEW

**Comment by backes** [comment]
Let me know if this should have a regression test somewhere.

File: src/wasm-delegations-fields.def [hunk 1]
```diff
@@ -461,12 +461,16 @@ switch (DELEGATE_ID) {
   }
   case Expression::Id::MemorySizeId: {
     DELEGATE_START(MemorySize);
+    DELEGATE_FIELD_TYPE(MemorySize, type);
```
Comment (kripken) [inline src/wasm-delegations-fields.def:464]: `type` is not needed in both cases, as it is present in all these classes and handled automatically.

**Review by kripken — COMMENTED** [review]
Thanks for the PR!

For a test you can use this, in `test/lit/passes/inlining_memory64.wat`:
```wat
;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.

;; RUN: foreach %s %t wasm-opt --inlining --enable-memory64 -S -o - | filecheck %s

(module
  ;; CHECK:      (type $none_=>_i64 (func (result i64)))

  ;; CHECK:      (memory $0 i64 256 256)
  (memory $0 i64 256 256)

  ;; CHECK:      (func $call-size (result i64)
  ;; CHECK-NEXT:  (block $__inlined_func$size (result i64)
  ;; CHECK-NEXT:   (memory.size)
  ;; CHECK-NEXT:  )
  ;; CHECK-NEXT: )
  (func $call-size (result i64)
    (call $size)
  )

  (func $size (result i64)
    ;; Upon inlining this code is copied, and as the memory is 64-bit the copied
    ;; instruction must remain 64-bit (and not have the default 32-bit size). If
    ;; we get that wrong then validation would fail here.
    (memory.size)
  )

  ;; CHECK:      (func $call-grow (result i64)
  ;; CHECK-NEXT:  (block $__inlined_func$grow (result i64)
  ;; CHECK-NEXT:   (memory.grow
  ;; CHECK-NEXT:    (i64.const 1)
  ;; CHECK-NEXT:   )
  ;; CHECK-NEXT:  )
  ;; CHECK-NEXT: )
  (func $call-grow (result i64)
    ;; As above, but for grow instead of size.
    (call $grow)
  )

  (func $grow (result i64)
    (memory.grow
      (i64.const 1)
    )
  )
)
```

File: src/wasm-delegations-fields.def (hunk 1 above)
Comment (backes) [reply src/wasm-delegations-fields.def]: Great, thanks!

**Comment by backes** [comment]
Thanks for providing the test! I remember that syntax from my LLVM days :)
Can you take another look please?

File: src/wasm-delegations-fields.def [hunk 2]
```diff
@@ -461,12 +461,14 @@ switch (DELEGATE_ID) {
...
   case Expression::Id::MemorySizeId: {
     DELEGATE_START(MemorySize);
+    DELEGATE_FIELD_TYPE(MemorySize, ptrType);
     DELEGATE_END(MemorySize);
     break;
   }
   case Expression::Id::MemoryGrowId: {
     DELEGATE_START(MemoryGrow);
     DELEGATE_FIELD_CHILD(MemoryGrow, delta);
+    DELEGATE_FIELD_TYPE(MemoryGrow, ptrType);
```
Comment (kripken) [inline src/wasm-delegations-fields.def:471]: The order here should be flipped (as the stack traversal pushes first), that is, the new line should be before `delta`.

Comment (kripken) [inline src/wasm-delegations-fields.def:471]:
```suggestion
    DELEGATE_FIELD_TYPE(MemoryGrow, ptrType);
    DELEGATE_FIELD_CHILD(MemoryGrow, delta);
```

Comment (kripken) [reply src/wasm-delegations-fields.def:471]: Oh, I guess github allows the reviewer to apply stuff like this, so I'll just do that.

**Review by kripken — APPROVED** [review]

**Pull request merged**